cdap-common/src/main/java/co/cask/cdap/gateway/handlers/StackHandler.java

```diff
@@ -0,0 +1,102 @@
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * Returns information about threads, including their stack traces.
+ */
+public class StackHandler extends AbstractHttpHandler {
+
+  private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
+
+  @Path(Constants.Gateway.API_VERSION_3 + "/system/services/{service-name}/stacks")
+  @GET
+  public void stacks(HttpRequest request, HttpResponder responder,
+                     @PathParam("service-name") String serviceName,
+                     @QueryParam("depth") @DefaultValue("20") int depth) {
+    StringWriter stringWriter = new StringWriter();
+    getThreadInfo(new PrintWriter(stringWriter), serviceName, depth);
+    responder.sendString(HttpResponseStatus.OK, stringWriter.toString());
+  }
+
+  /**
+   * Print all of the thread's information and stack traces.
+   */
+  private synchronized void getThreadInfo(PrintWriter stream, String serviceName, int depth) {
+    boolean contention = THREAD_MX_BEAN.isThreadContentionMonitoringEnabled();
+    long[] threadIds = THREAD_MX_BEAN.getAllThreadIds();
+    println(stream, "Process Thread Dump: %s", serviceName);
+    println(stream, "%s active threads", threadIds.length);
+    for (long tid : threadIds) {
+      ThreadInfo info = THREAD_MX_BEAN.getThreadInfo(tid, depth);
+      if (info == null) {
+        stream.println("  Inactive");
+        continue;
+      }
+      println(stream, "Thread %s",
+              getTaskName(info.getThreadId(), info.getThreadName()) + ":");
+      Thread.State state = info.getThreadState();
+      println(stream, "  State: %s", state);
+      println(stream, "  Blocked count: %s", info.getBlockedCount());
+      println(stream, "  Waited count: %s", info.getWaitedCount());
+      if (contention) {
+        println(stream, "  Blocked time: %s", info.getBlockedTime());
+        println(stream, "  Waited time: %s", info.getWaitedTime());
+      }
+      if (state == Thread.State.WAITING) {
+        println(stream, "  Waiting on %s", info.getLockName());
+      } else if (state == Thread.State.BLOCKED) {
+        println(stream, "  Blocked on %s", info.getLockName());
+        println(stream, "  Blocked by %s",
+                         getTaskName(info.getLockOwnerId(), info.getLockOwnerName()));
+      }
+      println(stream, "  Stack:");
+      for (StackTraceElement frame : info.getStackTrace()) {
+        println(stream, "    %s", frame.toString());
+      }
+    }
+    stream.flush();
+  }
+
+  private void println(PrintWriter stream, String format, Object... args) {
+    stream.printf(format, args).println();
```

Comment:
Seems unnecessary to have a one line method. You can do `stream.printf("Some stream %s\n", xyz);`. Pretty much like using System.out.

Reply:
But then we need to append printf's statement with "\n"?